### PR TITLE
add ingress endpoint

### PR DESCRIPTION
**Pull request by alaypatel07**

### CODE REVIEW

**Review by JaydipGabani — APPROVED** [review]
VISACK

File: state_transfer/endpoint/ingress/ingress.go [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+	return nil
+}
+
+func NewEndpoint(namespacedName types.NamespacedName, labels map[string]string) endpoint.Endpoint {
+	i := &IngressEndpoint{
+		namespacedName: namespacedName,
+		labels:         labels,
+		port:           6443,
+		// TODO: bring the subdomain as a param
+		hostname: namespacedName.Name + ".crane.dev",
```
Comment (djzager) [inline state_transfer/endpoint/ingress/ingress.go:155]: Is the using of the `subdomain` something you intend to include in this PR?

File: state_transfer/endpoint/ingress/ingress.go [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+			},
+			Selector: serviceSelector,
+			Type:     corev1.ServiceTypeNodePort,
+		},
+	}
+	// TODO: consider patching an existing object if it already exists
+	err := c.Create(context.TODO(), &service, &client.CreateOptions{})
+	if err != nil && !k8serrors.IsAlreadyExists(err) {
+		return err
+	}
```
[the remark this answers is not in the document]
Comment (alaypatel07) [reply state_transfer/endpoint/ingress/ingress.go:103]: https://github.com/backube/pvc-transfer/blob/02dbbbbb8c852936eb5d50125a128496f7e33cef/endpoint/loadbalancer/loadbalancer.go#L142 

@djzager when we migrate this to upstream, my plan was to use CreateOrUpdate. Everything in crane-lib for now unfortunately uses just Create, so for consistency can we continue to do this?

File: state_transfer/endpoint/ingress/ingress.go (hunk 1 above)
Comment (alaypatel07) [reply state_transfer/endpoint/ingress/ingress.go:155]: yes, will push an update. Gimme a few mins, good catch

**Review by djzager — COMMENTED** [review]
1 comment, otherwise this looks good to me

File: state_transfer/endpoint/ingress/ingress.go [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+	ingressPrefix := fmt.Sprintf("%s-%s", namespacedName.Name, namespacedName.Namespace)
+	if len(ingressPrefix) > 62 {
+		ingressPrefix = fmt.Sprintf("%s-%s", namespacedName.Name, getMD5Hash(namespacedName.Namespace))
+	}
+
+	i := &IngressEndpoint{
+		namespacedName: namespacedName,
+		labels:         labels,
+		port:           6443,
+		hostname:       ingressPrefix + "." + subdomain,
```
[the remark this answers is not in the document]
Comment (alaypatel07) [reply state_transfer/endpoint/ingress/ingress.go:166]: hmm, actually I should nuke the subdomain key from the struct, don't see it being used anywhere.

[the remark this answers is not in the document]
Comment (djzager) [reply state_transfer/endpoint/ingress/ingress.go:166]: that is an alternative approach :sunglasses:

[the remark this answers is not in the document]
Comment (alaypatel07) [reply state_transfer/endpoint/ingress/ingress.go:165]: done

**Review by sseago — APPROVED** [review]

**Review by djzager — APPROVED** [review]

**Pull request merged**